src/main/java/com/google/devtools/build/lib/hash/Blake3Hasher.java

```diff
@@ -0,0 +1,115 @@
+package com.google.devtools.build.lib.hash;
+
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class Blake3Hasher extends AbstractHasher {
+  // These constants match the native definitions in:
+  // https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3.h
+  public static final int KEY_LEN = 32;
+  public static final int OUT_LEN = 32;
+
+  // To reduce the number of calls made via JNI, buffer up to this many bytes.
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+
+  private final ReentrantLock rl = new ReentrantLock();
+  private ByteBuffer buffer;
+  private long hasher = -1;
+
+  public boolean isAllocated() {
+    return (hasher != -1);
+  }
+
+  private void initOnce() {
+    if (!isAllocated()) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+  }
+
+  public Blake3Hasher() {}
+
+  private void resetBuffer(int minLength) {
+    int length = Math.max(ONESHOT_THRESHOLD, minLength);
+    if (buffer == null || buffer.capacity() < length) {
+      buffer = ByteBuffer.allocateDirect(length);
+      buffer.order(ByteOrder.nativeOrder());
+    }
+    buffer.clear();
+  }
+
+  public void update(byte[] data, int offset, int length) {
+    rl.lock();
+
+    if (buffer == null) {
+      resetBuffer(length);
+    }
+
+    try {
+      if (buffer.remaining() < length) {
+        initOnce();
+        Blake3JNI.blake3_hasher_update(hasher, buffer, 0, buffer.position());
+        resetBuffer(length);
```

Comment:
something i worried about while poking at this was getting some single large chunk of data and then holding a buffer of that size for all time as a result; i ended up deciding that they can't possibly be so big that having 2x the buffer space of whatever our biggest read chunk is would ever matter.

this comment is not particularly actionable